### DISCUSSION TITLE
Looking for CRUD API Integration with Next

### DISCUSSION THREAD

**Discussion by pkellner** [vercel/next.js#22745] [Help]
Anyone tried to do an integration between json-server and next's CRUD API routes?  Having that work would allow for app unit testing easier without having to worry about json-server running on another port.  Also would make bundling tutorials easier there would be no second server or npm-run-all type hot mess to have to worry with.

(https://vercel.com/blog/10-next-js-tips-you-might-not-know#tip-7-crud-api-routes)

(https://www.npmjs.com/package/json-server#custom-routes-example)

**Comment by pkellner** [comment]
Continuing on this question (and train of thought), If I want to make my own simple CRUD server just for demo servers that would work consistently with either `npm run dev` or `npm start` after build (and I don't want to create a custom express server), where could I persist server type data when I do this?  That is, with json-server, it stores to a file, but that works because it's running a express server that has access to the local file system.  Is there something I could do that would be equivalent without a custom express server?

**Reply by borsemayur2** [reply]
I did the exact thing in [Aleph.js](alephjs.org/) (Deno alternative for Next.js) using **Deno** (readTextFile, writeTextFile) methods. I think the same should be achievable in Next.js. I used a JSON file to read/write data.

**Reply by pkellner** [reply]
@borsemayur2 , Deno is on my list. Did you post the code anyplace for that? I'm still wrapping my head around the API routes keeping track of what is running in the client and what is on the server.  I'm not yet one with that but hope to be someday.

**Reply by borsemayur2** [reply]
Actually, I wrote it for my upcoming book on Deno so it can't be made public yet but, I can create a min repro repo that you can check. I'll also try to achieve the same in Next.js. I'll share it once completed.

**Reply by pkellner** [reply]
Hi @borsemayur2 , have you had a chance to push a small example someplace?

**Reply by borsemayur2** [reply]
I've quickly created an Aleph JSON server example and have invited you as a collaborator. Please message me if you have any queries or need more details. I'll try to create the same in Next.js (node) soon.

**Reply by pkellner** [reply]
Thanks, just accepted it and taking a look. I've been wrestling with making one myself all morning so this should help.  I'm assuming this will only work in dev mode, as it seems the fs package with nextjs, according to stackoverflow, only works dev time for writing.

**Reply by pkellner** [reply]
and @borsemayur2 , if you are having a book launch party, put me on your invite list!

**Comment by pkellner** [comment]
Is there a better place to post questions that use to get some attention at least (like the old Spectrum forum)?

**Reply by leerob** [reply]
Hey! Following up from your message on Twitter too, this is the place: https://twitter.com/leeerob/status/1369813972233420800

I haven't tried to use `json-server` before. If you're using Prisma, this package might be helpful: https://next-crud.js.org/

**Reply by leerob** [reply]
When you spin up Prisma, you can use a simple SQLite DB locally, which could work nicely.

**Reply by pkellner** [reply]
@leerob thanks, I'm really looking something I can bundle in a simple "npm run dev" and "npm start" type scenario for demos without much in the way of external dependencies.  People often run my courseware in stripped down linux servers that run npm, but having any kind of dependencies on persistent stores is always a pain (as they often map ports and other nefarious things to make it "easier").

**Reply by borsemayur2** [reply]
Agree. I think a local JSON file as a database serves the purpose of learning and demos.

**Reply by pkellner** [reply]
That looks interesting.  Just looking at Prisma.  Looks a lot simple than I had thought previously.  The CRUD helper may work for me. 

> Hey! Following up from your message on Twitter too, this is the place: https://twitter.com/leeerob/status/1369813972233420800
> 
> 
> 
> I haven't tried to use `json-server` before. If you're using Prisma, this package might be helpful: https://next-crud.js.org/
